### PR TITLE
fix(k8s)!: support k8s multi container

### PR DESCRIPTION
**Pull request by smtan-gl**
## Description
When using Trivy k8s, the current report metadata field for workloads with multiple containers only includes metadata from one image.
This update changes the metadata field to an array, allowing it to capture metadata from all the scanned images within the workload.

This is a draft PR to gather feedback on whether this approach is acceptable, and I will update the tests accordingly thereafter.

<details>
<summary>Report before change</summary>

```
{
  "ClusterName": "",
  "Findings": [
    {
      "Namespace": "default",
      "Kind": "Pod",
      "Name": "nginx-fluentd-pod",
      "Metadata": {
        "OS": {
          "Family": "debian",
          "Name": "12.6"
        },
          "RepoTags": [
            "nginx:latest"
          ],
          "DiffIDs": []
      },
      "Results": [
        {
          "Target": "nginx:latest (debian 12.6)",
          "Class": "os-pkgs",
          "Type": "debian",
          "Packages": [],
          "Vulnerabilities": []
        },
        {
          "Target": "fluent/fluentd:v1.17-armhf-debian (debian 12.6)",
          "Class": "os-pkgs",
          "Type": "debian",
          "Packages": [],
          "Vulnerabilities": []
        },
        {
          "Target": "Ruby",
          "Class": "lang-pkgs",
          "Type": "gemspec",
          "Packages": [],
          "Vulnerabilities": []
        }
      ]
    }
  ]
}
```
</details>

<details>
<summary>Report after change</summary>

```
{
  "ClusterName": "",
  "Findings": [
    {
      "Namespace": "default",
      "Kind": "Pod",
      "Name": "nginx-fluentd-pod",
      "Metadata": [
        {
          "OS": {
            "Family": "debian",
            "Name": "12.6"
          },
          "RepoTags": [
            "nginx:latest"
          ],
          "DiffIDs": []
        },
        {
          "OS": {
            "Family": "debian",
            "Name": "12.6"
          },
          "RepoTags": [
            "fluent/fluentd:v1.17-armhf-debian"
          ],
          "DiffIDs": []
        }
      ],
      "Results": [
        {
          "Target": "nginx:latest (debian 12.6)",
          "Class": "os-pkgs",
          "Type": "debian",
          "Packages": [],
          "Vulnerabilities": []
        },
        {
          "Target": "fluent/fluentd:v1.17-armhf-debian (debian 12.6)",
          "Class": "os-pkgs",
          "Type": "debian",
          "Packages": [],
          "Vulnerabilities": []
        },
        {
          "Target": "Ruby",
          "Class": "lang-pkgs",
          "Type": "gemspec",
          "Packages": [],
          "Vulnerabilities": []
        }
      ]
    }
  ]
}

```
</details>

## Related issues
- Close [Unexpected JSON output when --report summary is set in trivy k8s for multi-container workloads](https://github.com/aquasecurity/trivy/issues/5889)

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/k8s/report/report.go
Comment (afdesk) [inline pkg/k8s/report/report.go:68]: can we update `Resource`? I think it's a bit hard to keep 2 similar structs.

Comment (afdesk) [reply pkg/k8s/report/report.go:68]: also there are no tests for this case.

**Review by afdesk — CHANGES_REQUESTED** [review]

**Comment by afdesk** [comment]
Hi @smtan-gl  thanks for the contribution! it's really nice and important.

I left some comments about this PR. 

also I took a look at #5889 
it seems we have to add more tests for these cases...

File: pkg/k8s/report/report.go
Comment (smtan-gl) [reply pkg/k8s/report/report.go:68]: Thanks for the feedback, I'm not sure why I assumed the `Resource` struct was used by other packages in Trivy, but it doesn't seem to be the case. I'll work on updating the `Resource` struct accordingly.

> also there are no tests for this case.
Regarding the tests, I wanted to get feedback on the approach before proceeding. I'll focus on adding the tests next.

**Comment by smtan-gl** [comment]
> Hi @smtan-gl thanks for the contribution! it's really nice and important.
> 
> I left some comments about this PR.
> 
> also I took a look at #5889 it seems we have to add more tests for these cases...

@afdesk any chance you might have context on this [question](https://github.com/aquasecurity/trivy/issues/5889#issuecomment-2330536466)?

**Comment by afdesk** [comment]
> @afdesk any chance you might have context on this [question](https://github.com/aquasecurity/trivy/issues/5889#issuecomment-2330536466)?

If I understand correctly there is an option to create an image with several tags, so it should be a slice.
https://stackoverflow.com/questions/31963525/is-it-possible-for-image-to-have-multiple-tags

**Comment by smtan-gl** [comment]
@afdesk I've updated the PR, could you have another look please 😄

**Comment by afdesk** [comment]
> @afdesk I've updated the PR, could you have another look please 😄

sure, thanks! I'll take a look

**Comment by afdesk** [comment]
@smtan-gl could you sign CLA for a while? thanks!

**Comment by smtan-gl** [comment]
> @smtan-gl could you sign CLA for a while? thanks!

@afdesk, I've just signed it. Thank you.

**Review by afdesk — CHANGES_REQUESTED** [review]
Thanks for your contribute! I left a few comments

**Comment by afdesk** [comment]
it's a BREAKING change, so we have to create a new discussion about it.
like here: https://github.com/aquasecurity/trivy/discussions/7010
we'll do it after the PR is merged. just to remember about it

**Comment by afdesk** [comment]
@smtan-gl  also, I'm not sure that this PR fixes #5889...

**Comment by smtan-gl** [comment]
@afdesk, the bug https://github.com/aquasecurity/trivy/issues/5889 occurs only with the `--report summary` configuration. 

In your example, you used `--report all` which [does not consolidate the report](https://github.com/aquasecurity/trivy/blob/35c4262d0b5ac18376a2ca53213b476103b5d51a/pkg/k8s/report/json.go#L23). As such, there are 2 reports in the `Resources` field, 1 for each of the scanned images containing only 1 metadata.

When I run the command using `--report summary`: 
```
$ ./tr k8s --report summary -f json -o res2.json --include-namespaces k8s-test --timeout 30m --debug --include-kinds Pod --scanners vuln
```

It [consolidates the reports](https://github.com/aquasecurity/trivy/blob/35c4262d0b5ac18376a2ca53213b476103b5d51a/pkg/k8s/report/json.go#L28) into 1 consolidated report in the `Findings` field, and you can see that there are 2 metadata for the 2 images.

Screenshot 2024-09-19 at 2 02 05 PM

File: pkg/k8s/report/report_test.go
Comment (smtan-gl) [reply pkg/k8s/report/report_test.go]: @afdesk that's possible. I've [updated the code](https://github.com/aquasecurity/trivy/pull/7444/commits/ed599f1ae16da89793bc0c0a172a81706e0b6c85) to use `assert.Equal` and consolidated it in `TestReport_consolidate`.

File: pkg/k8s/report/report_test.go
Comment (smtan-gl) [reply pkg/k8s/report/report_test.go]: @afdesk that's possible. I've made the change [here](https://github.com/aquasecurity/trivy/pull/7444/commits/ed599f1ae16da89793bc0c0a172a81706e0b6c85).

**Review by afdesk — APPROVED** [review]
LGTM!

File: pkg/k8s/report/report_test.go
Comment (afdesk) [reply pkg/k8s/report/report_test.go]: great! thanks!

**Comment by afdesk** [comment]
> @afdesk, the bug #5889 occurs only with the `--report summary` configuration.

exactly! thanks for clarify!

**Comment by smtan-gl** [comment]
Thanks for reviewing the PR @afdesk 😄 May I know what's the next steps needed to have this PR merged?

**Comment by afdesk** [comment]
> Thanks for reviewing the PR @afdesk 😄 May I know what's the next steps needed to have this PR merged?

@smtan-gl thanks for your contribution again. 
It's a breaking change so let's wait feedback from another maintainers and community

**Comment by smtan-gl** [comment]
> > Thanks for reviewing the PR @afdesk 😄 May I know what's the next steps needed to have this PR merged?
> 
> @smtan-gl thanks for your contribution again. It's a breaking change so let's wait feedback from another maintainers and community

Hi @afdesk would you have an estimate on how long we need to wait for feedback?

**Comment by afdesk** [comment]
Hi @smtan-gl 

there is a concern about this solution: how people associate the result with the metadata if it's N x N?

wdyt about it? thanks!

**Comment by smtan-gl** [comment]
> Hi @smtan-gl
> 
> there is a concern about this solution: how people associate the result with the metadata if it's N x N?
> 
> wdyt about it? thanks!

Thanks for the response @afdesk. I thought we could match the vulnerability's `Layer.DiffID` with the metadata `DiffIDs`, but I've realized that the `DiffIDs` across multiple metadata entries might not be unique. Would it make sense to introduce an `ImageID` field to vulnerability?

For your convenience, I've attached below a JSON report from a scan with changes from this MR.

[multi-image-pod.json](https://github.com/user-attachments/files/17129086/multi-image-pod.json)

**Comment by afdesk** [comment]
@smtan-gl there are a few ideas.

1. @knqyf263 suggests to have Metadata and Results per image. 
<details>
<summary>a sample json for a new field</summary>

```json
  "Images": [
    {
      "Metadata": {
        "OS": {
          "Family": "debian",
          "Name": "12.6"
        },
        "RepoTags": [
          "image-a:latest"
        ]
      },
      "Results": [
        {
          "Target": "image-a:latest(debian 12.6)",
          "Class": "os-pkgs",
          "Type": "debian",
          "Vulnerabilities": []
        }
      ]
    },
    {
      "Metadata": {
        "OS": {
          "Family": "ubuntu",
          "Name": "22.04"
        },
        "RepoTags": [
          "image-b:latest"
        ]
      },
      "Results": [
        {
          "Target": "image-b:latest(ubuntu 22.04)",
          "Class": "os-pkgs",
          "Type": "ubuntu",
          "Vulnerabilities": []
        }
      ]
    }
  ]
```
</details>

Next ideas came from the point `summary` report basically is for reference and for table output.

2. disable JSON format for summary reports.
People can summarize results by themselves. and maybe someone will create a plugin for it.

3. Pods are the smallest deployable units of computing that you can create and manage in Kubernetes. so If we say Pod is a single object, and we don't distinguish image. and then the current approach is fine.

Wdyt @smtan-gl ?

**Comment by smtan-gl** [comment]
> Next ideas came from the point summary report basically is for reference and for table output.

Thanks for the response @afdesk. I agree with your points. However, I assume that there are other users who already rely on summary report in JSON format, so disabling it might affect them. For that reason, I support moving forward with approach 3.

**Comment by afdesk** [comment]
@smtan-gl just small update that I didn't forget about your job.
I hope we'll resolve it before next release. 
thanks again

**Pull request merged**
